src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -59,5 +63,205 @@ public void TestSignaturesV2_3_X()
+            TestAndValidateForeach(
+                streamInputPath,
+                new TestForeachWriter(),
+                false,
+                3,
+                3,
+                0,
+                Enumerable.Range(101, 99));
+
+            // Test scenario where IForeachWriter.Open returns false.
+            // When IForeachWriter.Open returns false, then IForeachWriter.Process
+            // is not called.  Verify that:
+            // - Triggering stream will not throw an exception
+            // - 3 CSV files will be created in the temporary directory.
+            // - 0 Exception files will be created in the temporary directory.
+            // - The CSV files will not contain valid data to read.
+            TestAndValidateForeach(
+                streamInputPath,
+                new TestForeachWriterOpenFailure(),
+                false,
+                3,
+                3,
+                0,
+                Enumerable.Empty<int>());
+
+
+            // Test scenario where ForeachWriter.Process throws an Exception.
+            // When IForeachWriter.Process throws an Exception, then the exception
+            // is rethrown by ForeachWriterWrapper. We will limit the partitions
+            // to 1 to make validating this scenario simpler. Verify that:
+            // - Triggering stream throws an exception.
+            // - 1 CSV file will be created in the temporary directory.
+            // - 1 Exception will be created in the temporary directory. The
+            //   thrown exception from Process() will be sent to Close().
+            // - The CSV file will not contain valid data to read.
+            TestAndValidateForeach(
+                    streamInputPath,
+                    new TestForeachWriterProcessFailure(),
+                    true,
+                    1,
+                    1,
+                    1,
+                    Enumerable.Empty<int>());
+        }
+
+        private void TestAndValidateForeach(
+            string streamInputPath,
+            TestForeachWriter foreachWriter,
+            bool foreachThrows,
+            int partitions,
+            long expectedCSVFiles,
+            long expectedExceptionFiles,
+            IEnumerable<int> expectedOutput)
+        {
+            // Temporary folder the TestForeachWriter will write to.
+            using var dstTempDirectory = new TemporaryDirectory();
+
+            foreachWriter.WritePath = dstTempDirectory.Path;
+
+            // Read streamInputPath, repartitions data into `partitions`, then
+            // calls TestForeachWriter on the data.
+            DataStreamWriter dsw = _spark
+                .ReadStream()
+                .Schema("id INT")
+                .Json(streamInputPath)
+                .Repartition(partitions)
+                .WriteStream()
+                .Foreach(foreachWriter);
+
+            // Trigger the stream batch once.
+            if (foreachThrows)
```

Comment:
looks like you can do `if (expectedExceptionFiles > 0)`?

Reply:
Yup!